File: r/sedonadb/R/context.R
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

#' Create a DataFrame from one or more Parquet files
#'
#' The query will only be executed when requested.
#'
#' @param path One or more paths or URIs to Parquet files
#'
#' @returns A sedonadb_dataframe
#' @export
#'
#' @examples
#' path <- system.file("files/natural-earth_cities_geo.parquet", package = "sedonadb")
#' sd_read_parquet(path) |> head(5) |> sd_preview()
#'
sd_read_parquet <- function(path) {
  ctx <- ctx()
  df <- ctx$read_parquet(path)
  new_sedonadb_dataframe(ctx, df)
}

#' Create a DataFrame from SQL
#'
#' The query will only be executed when requested.
#'
#' @param sql A SQL string to execute
#'
#' @returns A sedonadb_dataframe
#' @export
#'
#' @examples
#' sd_sql("SELECT ST_Point(0, 1) as geom") |> sd_preview()
#'
sd_sql <- function(sql) {
  ctx <- ctx()
  df <- ctx$sql(sql)
  new_sedonadb_dataframe(ctx, df)
}

#' Create or Drop a named view
#'
#' Remove a view created with [sd_to_view()] from the context.
#'
#' @param table_ref The name of the view reference
#' @returns The context, invisibly
#' @export
#'
#' @examples
#' sd_sql("SELECT 1 as one") |> sd_to_view("foofy")
#' sd_view("foofy")
#' sd_drop_view("foofy")
#' try(sd_view("foofy"))
#'
sd_drop_view <- function(table_ref) {
  ctx <- ctx()
  ctx$deregister_table(table_ref)
  invisible(ctx)
}

#' @rdname sd_drop_view
#' @export
sd_view <- function(table_ref) {
  ctx <- ctx()
  df <- ctx$view(table_ref)
  new_sedonadb_dataframe(ctx, df)
}

#' Register a user-defined function
#'
#' Several types of user-defined functions can be registered into a session
#' context. Currently, the only implemented variety is an external pointer
#' to a Rust `FFI_ScalarUDF`, an example of which is available from the
#' [DataFusion Python documentation](https://github.com/apache/datafusion-python/blob/6f3b1cab75cfaa0cdf914f9b6fa023cb9afccd7d/examples/datafusion-ffi-example/src/scalar_udf.rs).
#'
#' @param udf An object of class 'datafusion_scalar_udf'
#'
#' @returns NULL, invisibly
#' @export
#'
sd_register_udf <- function(udf) {
  ctx <- ctx()
  ctx$register_scalar_udf(udf)
}

# We use just one context for now. In theory we could support multiple
# contexts with a shared runtime, which would scope the registration
# of various components more cleanly from the runtime.
ctx <- function() {
  if (is.null(global_ctx$ctx)) {
    global_ctx$ctx <- InternalContext$new()
  }

  global_ctx$ctx
}

global_ctx <- new.env(parent = emptyenv())
global_ctx$ctx <- NULL



#' Configure PROJ
#'
#' Performs a runtime configuration of PROJ, which can be used in place of
#' a build-time linked version of PROJ or to add in support if PROJ was
#' not linked at build time.
#'
#' @param preset One of:
#'   - `"homebrew"`: Look for PROJ installed by Homebrew. This is the easiest
#'     option on MacOS.
#'   - `"system"`: Look for PROJ in the platform library load path (e.g.,
#'     after installing system proj on Linux).
#'   - `"auto"`: Try all presets in the order listed above, issuing a warning
#'     if none can be configured.
#' @param shared_library An absolute or relative path to a shared library
#'   valid for the platform.
#' @param database_path A path to proj.db
#' @param search_path A path to the data files required by PROJ for some
#'   transforms.
#'
#' @returns NULL, invisibly
#' @export
#'
#' @examples
#' sd_configure_proj("auto")
#'
sd_configure_proj <- function(preset = NULL,
                              shared_library = NULL,
                             database_path = NULL,
                             search_path = NULL) {
  if (!is.null(preset)) {
    switch (preset,
      homebrew = {
        configure_proj_prefix(Sys.getenv("HOMEBREW_PREFIX", "/opt/homebrew"))
        return(invisible(NULL))
      },
      system = {
        configure_proj_system()
        return(invisible(NULL))
      },
      auto = {
        presets <- c("homebrew", "system")
        errors <- c()
        for (preset in presets) {
          maybe_err <- try(sd_configure_proj(preset), silent = TRUE)
          if (!inherits(maybe_err, "try-error")) {
            return(invisible(NULL))
          } else {
            errors <- c(errors, sprintf("%s: %s", preset, maybe_err))
          }
        }

        packageStartupMessage(
          sprintf(
            "Failed to configure PROJ (tried %s):\n%s",
            paste0("'", presets, "'", collapse = ", "),
            paste0(errors, collapse = "\n")
          )
        )

        return(invisible(NULL))
      },
      stop(sprintf("Unknown preset: '%s'", preset))
    )
  }

  # We could check a shared library with dyn.load(), but this may error for
  # valid system PROJ that isn't an absolute filename.

  if (!is.null(database_path)) {
    if (!file.exists(database_path)) {
      stop(sprintf("Invalid database path: '%s' does not exist", database_path))
    }
  }

  if (!is.null(search_path)) {
    if (!dir.exists(search_path)) {
      stop(sprintf("Invalid search path: '%s' does not exist", search_path))
    }
  }

  configure_proj_shared(
    shared_library_path = shared_library,
    database_path = database_path,
    search_path = search_path
  )
}

configure_proj_system <- function() {
  sd_configure_proj(shared_library = proj_dll_name())
}

configure_proj_prefix <- function(prefix) {
  if (!dir.exists(prefix)) {
    stop(sprintf("Can't configure PROJ from prefix '%s': does not exist", prefix))
  }

  sd_configure_proj(
    shared_library = file.path(prefix, "lib", proj_dll_name()),
    database_path = file.path(prefix, "share", "proj", "proj.db"),
    search_path = file.path(prefix, "share", "proj")
  )
}

proj_dll_name <- function() {
  switch(tolower(Sys.info()[["sysname"]]),
    windows = "proj.dll",
    darwin = "libproj.dylib",
    linux = "libproj.so",
    stop(sprintf("Can't determine system PROJ shared library name for OS: %s", Sys.info()[["sysname"]]))
  )
}

File: r/sedonadb/src/rust/src/context.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
use std::sync::Arc;

use arrow_array::RecordBatchReader;
use arrow_schema::ArrowError;
use datafusion::catalog::{MemTable, TableProvider};
use datafusion_ffi::udf::FFI_ScalarUDF;
use savvy::{savvy, savvy_err, IntoExtPtrSexp, Result};

use sedona::{context::SedonaContext, record_batch_reader_provider::RecordBatchReaderProvider};
use sedona_geoparquet::provider::GeoParquetReadOptions;
use tokio::runtime::Runtime;

use crate::{
    dataframe::{new_data_frame, InternalDataFrame},
    ffi::{import_array_stream, import_scalar_udf, import_table_provider, FFIScalarUdfR},
    runtime::wait_for_future_captured_r,
};

#[savvy]
pub struct InternalContext {
    pub inner: Arc<SedonaContext>,
    pub runtime: Arc<Runtime>,
}

#[savvy]
impl InternalContext {
    pub fn new() -> Result<Self> {
        let runtime = tokio::runtime::Builder::new_multi_thread()
            .enable_all()
            .build()?;

        let inner = wait_for_future_captured_r(&runtime, SedonaContext::new_local_interactive())??;

        Ok(Self {
            inner: Arc::new(inner),
            runtime: Arc::new(runtime),
        })
    }

    pub fn read_parquet(&self, paths: savvy::Sexp) -> Result<InternalDataFrame> {
        let paths_strsxp = savvy::StringSexp::try_from(paths)?;
        let table_paths = paths_strsxp
            .iter()
            .map(|s| s.to_string())
            .collect::<Vec<_>>();

        let inner_context = self.inner.clone();
        let inner = wait_for_future_captured_r(&self.runtime, async move {
            inner_context
                .read_parquet(table_paths, GeoParquetReadOptions::default())
                .await
        })??;

        Ok(new_data_frame(inner, self.runtime.clone()))
    }

    pub fn sql(&self, query: &str) -> Result<InternalDataFrame> {
        let query_string = query.to_string();
        let inner_context = self.inner.clone();
        let inner = wait_for_future_captured_r(&self.runtime, async move {
            inner_context.sql(&query_string).await
        })??;
        Ok(new_data_frame(inner, self.runtime.clone()))
    }

    pub fn view(&self, table_ref: &str) -> Result<InternalDataFrame> {
        let inner_context = self.inner.clone();
        let table_ref_string = table_ref.to_string();
        let inner = wait_for_future_captured_r(&self.runtime, async move {
            inner_context.ctx.table(table_ref_string).await
        })??;
        Ok(new_data_frame(inner, self.runtime.clone()))
    }

    pub fn data_frame_from_array_stream(
        &self,
        stream_xptr: savvy::Sexp,
        collect_now: bool,
    ) -> savvy::Result<InternalDataFrame> {
        let stream_reader = import_array_stream(stream_xptr)?;

        // Some readers are sensitive to being collected on the R thread or not, so
        // provide the option to collect everything immediately.
        let provider: Arc<dyn TableProvider> = if collect_now {
            let schema = stream_reader.schema();
            let batches = stream_reader.collect::<std::result::Result<Vec<_>, ArrowError>>()?;
            Arc::new(MemTable::try_new(schema, vec![batches])?)
        } else {
            Arc::new(RecordBatchReaderProvider::new(Box::new(stream_reader)))
        };

        let inner = self.inner.ctx.read_table(provider)?;
        Ok(new_data_frame(inner, self.runtime.clone()))
    }

    pub fn data_frame_from_table_provider(
        &self,
        provider_xptr: savvy::Sexp,
    ) -> Result<InternalDataFrame> {
        let provider = import_table_provider(provider_xptr)?;
        let inner = self.inner.ctx.read_table(provider)?;
        Ok(new_data_frame(inner, self.runtime.clone()))
    }

    pub fn deregister_table(&self, table_ref: &str) -> savvy::Result<()> {
        self.inner.ctx.deregister_table(table_ref)?;
        Ok(())
    }

    pub fn scalar_udf_xptr(&self, name: &str) -> savvy::Result<savvy::Sexp> {
        if let Some(udf) = self.inner.ctx.state().scalar_functions().get(name) {
            let ffi_scalar_udf: FFI_ScalarUDF = udf.clone().into();
            let mut ffi_xptr = FFIScalarUdfR(ffi_scalar_udf).into_external_pointer();
            unsafe { savvy_ffi::Rf_protect(ffi_xptr.0) };
            ffi_xptr.set_class(vec!["datafusion_scalar_udf"])?;
            unsafe { savvy_ffi::Rf_unprotect(1) };

            Ok(ffi_xptr)
        } else {
            Err(savvy_err!("Scalar UDF '{name}' was not found"))
        }
    }

    pub fn register_scalar_udf(&self, scalar_udf_xptr: savvy::Sexp) -> savvy::Result<()> {
        let scalar_udf = import_scalar_udf(scalar_udf_xptr)?;
        self.inner.ctx.register_udf(scalar_udf);
        Ok(())
    }
}
